client/core/core.go

```diff
@@ -7437,3 +7479,223 @@ func (c *Core) findActiveOrder(oid order.OrderID) (*trackedTrade, error) {
+				tick.Stop()
+				return
+			case <-tick.C:
+				c.refreshFiatRates(ctx)
+			}
+			tick = c.nextRateSourceTick()
+		}
+	}()
+}
+
+// refreshFiatRates refreshes the fiat rates for rate sources whose values have
+// not been updated since fiatRateRequestInterval.
+func (c *Core) refreshFiatRates(ctx context.Context) {
+	ctx, cancel := context.WithTimeout(ctx, 4*time.Second)
+	defer cancel()
+
+	var wg sync.WaitGroup
+	supportedAssets := c.SupportedAssets()
+	timeNow := time.Now()
+	c.ratesMtx.Lock()
+	for _, source := range c.fiatRateSources {
+		if time.Since(source.lastRequest) > fiatRateRequestInterval {
+			wg.Add(1)
+			go func(source *commonRateSource) {
+				source.lastRequest = timeNow
+				defer wg.Done()
+				source.refreshRates(ctx, c.log, supportedAssets)
+			}(source)
+		}
+	}
+	c.ratesMtx.Unlock()
+	wg.Wait()
+
+	fiatRatesMap := c.fiatConversions()
+	if len(fiatRatesMap) != 0 {
+		c.notify(newFiatRatesUpdate(fiatRatesMap))
+	}
+}
+
+// nextRateSourceTick checks the fiat rate source's last request, and calculates
+// when the next cycle should run.
+func (c *Core) nextRateSourceTick() *time.Timer {
+	c.ratesMtx.RLock()
+	defer c.ratesMtx.RUnlock()
+
+	minTick := 10 * time.Second
+	tOldest := time.Now()
+	for _, source := range c.fiatRateSources {
+		t := source.lastRequest
+		if t.Before(tOldest) {
+			tOldest = t
+		}
+	}
+	tilNext := fiatRateRequestInterval - time.Since(tOldest)
+	if tilNext < minTick {
+		tilNext = minTick
+	}
+	return time.NewTimer(tilNext)
+}
+
+// FiatRateSources returns a list of fiat rate sources and their individual
+// status.
+func (c *Core) FiatRateSources() map[string]bool {
+	c.ratesMtx.RLock()
+	defer c.ratesMtx.RUnlock()
+	rateSources := make(map[string]bool, len(fiatRateFetchers))
+	for token := range fiatRateFetchers {
+		rateSources[token] = c.fiatRateSources[token] != nil
+	}
+	return rateSources
+}
+
+// fiatConversions returns fiat rate for all supported assets that have a
+// wallet. It also checks if fiat rates are expired and does some clean-up.
+func (c *Core) fiatConversions() map[uint32]float64 {
+	supportedAssets := c.SupportedAssets()
+
+	c.ratesMtx.RLock()
+	defer c.ratesMtx.RUnlock()
+	disabledCount := 0
+	fiatRatesMap := make(map[uint32]float64, len(supportedAssets))
+	for assetID := range supportedAssets {
+		var rateSum float64
+		var sources int32
+		for token, source := range c.fiatRateSources {
+			// Remove fiat rate source with expired exchange rate data.
+			if source.isExpired(fiatRateDataExpiry) {
+				delete(c.fiatRateSources, token)
+				disabledCount++
+				continue
+			}
+			rateInfo := source.assetRate(assetID)
+			if rateInfo != nil && time.Since(rateInfo.lastUpdate) < fiatRateDataExpiry {
+				sources++
+				rateSum += rateInfo.rate
+			}
+		}
+		if rateSum != 0 {
+			fiatRatesMap[assetID] = rateSum / float64(sources) // get average rate.
+		}
+	}
+
+	// Ensure disabled fiat rate fetchers are saved to database.
+	if disabledCount > 0 {
+		c.saveDisabledRateSources()
```

Comment:
This should have a WRN log I think.  It means something is failing, although hopefully remotely, and the user's source selections were changed for them.

Reply:
Philosophically, pruning dead rate sources is a job only for `refreshFiatRates` (after it completes `wg.Wait`).
`fiatConversions()` is primarily a getter method, called from the `(*Core).User` method that is hit quite frequently by external consumers and should generally by fast and not modify state.